### ISSUE TITLE
Compiler doesn't understand returned arrows

### ISSUE THREAD

**Issue by sebdelvalle** [Microsoft/TypeScript#16828]
**TypeScript Version:**  2.4.0 / nightly (2.5.0-dev.201xxxxx)

**Code**

```ts
let a = a => `'${a?a:''}'`
console.log(a())
```

**Expected behavior:**

Shouldn't throw any error. It works fine in Chrome and Firefox console, and in Flow.

**Actual behavior:**

throws:
```
Supplied parameters do not match any signature of call target.
```

Edit: changed my `Date` example since it would fail with empty dates.

**Comment by kitsonk** [comment]
You are indicating, the way you have written it, that `x` is not an optional argument.  If you want `x` to be optional you have to write it this way:

```ts
let a = (x?) => new Date(x)
console.log(a())
```

Now you have silly code that does not produce errors.

**Comment by sebdelvalle** [comment]
We're talking about regular ES6, which TypeScript should support out of the box. Just type my code in your browser's console and you will see it complies with the standard and runs fine.

**Comment by poseidonCore** [comment]
This  transpiles perfectly in TS playground (TypeScript 2.4.1?):

``
let a = a => `'${a ? a : ''}'`;
``

goes to 

``
var a = function (a) { return "'" + (a ? a : '') + "'"; };
``

**Comment by sebdelvalle** [comment]
The definition works fine. The problem is when you call it, the compiler throws the mentioned error.

**Comment by kitsonk** [comment]
```ts
function a(x) { return new Date(x) }

a(); // Supplied parameters do not match any signature of call target.
```

It has nothing to do with the arrow functions.  TypeScript does not have a `--stupid` compiler flag, I am afraid.  If you specify an argument, and it is not marked as optional, TypeScript enforces it.  You can feel free to ignore these errors, as it does not effect the emit.

Some would argue that Facebooks Flow is operating very unsoundly.

**Comment by kitsonk** [comment]
And since you edited your first example of code:

```ts
function a(a) { return `'${a ? a : ''}'` }
console.log(a()) // Supplied parameters do not match any signature of call target.
```

**Comment by jcalz** [comment]
The fix for the edited code is:
```ts
let a = (a?) => `'${a?a:''}'`
console.log(a()) // no error
```

**Comment by kitsonk** [comment]
Yes, which is the way I [_fixed_](https://github.com/Microsoft/TypeScript/issues/16828#issuecomment-311998084) his original example.

**Comment by RyanCavanaugh** [comment]
Flow's guessing at what makes a parameter optional is rather suspect IMO. If you want an optional parameter, use `?`. If you don't, don't.
```js
var fn = a => {
  if (a === undefined) throw new Error('Bad argument count!');
  return a;
};

fn(); // Flow says it's OK
```
<sub>(in before "but that code is TECHNICALLY VALID maybe you wanted to throw an exception!")</sub>

> We're talking about regular ES6, which TypeScript should support out of the box. Just type my code in your browser's console and you will see it runs fine.

Please read https://stackoverflow.com/questions/41750390/what-does-all-legal-javascript-is-legal-typescript-mean
